### PR TITLE
Parse everything as a block and simplify the output

### PR DESCRIPTION
**Pull request by youknowriad**
This PR seeks to simplifie the grammar to match the requirements of the Multi-Instance Gutenberg prototype.

The principles here are:
 - Everything is a block (even text nodes)
 - Avoid parsing HTML nodes (fixes a lot of issues, invalid html etc...)
 - Simplify the block object { blockType, rawContent, attrs }

### CODE REVIEW

**Comment by dmsnell** [comment]
thanks for the work @youknowriad - at this point the grammar isn't well developed primarily because it hasn't needed to be.

I'd caution against some of these changes like not parsing HTML and only accepting balanced blocks because if we do that we will lose the ability to describe partial posts, meaning we will be preemptively hurting any collaborative efforts and quite possibly introduce quirks into the system unintentionally…

in other words, there is value in having an _invalid_ document but a _valid_ parse

**Comment by youknowriad** [comment]
> in other words, there is value in having an invalid document but a valid parse

We currently have an invalid parse if we have an invalid document. unclosed HTML tags like `` are not working properly.

But more importantly, while working with the grammar on the `per-block` prototype, I did not find any value in parsing the HTML content of a block, since it's mostly a "serialization" of the state. On the HTML blocks, we just use it as the content of TinyMCE and in other more "controlled" blocks like images, paragraphs,embeds ... we ignore it completely.

**Comment by dmsnell** [comment]
> We currently have an invalid parse if we have an invalid document. unclosed HTML tags like  are not working properly.

yeah. that's a problem with the grammar implementation though and not a design choice.

> I did not find any value in parsing the HTML content of a block, since it's mostly a "serialization" of the state

this is part of the reason the grammar wasn't developed more - wasting effort building a thing that may not be the right thing.

I'd still suggest that there's value there though when we think of partial updates and synchronized updates. HTML is a tree but text is a list. storing the contents as a list of tokens is a design choice to make the validation separate from the parse. I would still like to see us be able to identify operations which make a document invalid, provide a meaningful alert to the author, and provide an undo operation. these things I think depend on parsing the structure of the contents vs just leaving them as a black box

**Comment by youknowriad** [comment]
> I would still like to see us be able to identify operations which make a document invalid, provide a meaningful alert to the author, and provide an undo operation. these things I think depend on parsing the structure of the contents vs just leaving them as a black box

There are some discussions currently on this on the gutenberg repo and we're seeing two approaches here:

 - The one saying that we should not care that much about invalid HTML since we're building an editor which should diminish drastically the need to edit the HTML directly where all the block state is stored in the attributes and/or post meta. Basically, we don't care about the HTML, it's just the output. (I think @aduth suggested this approach at first)

 - The other approach I used in the `per-block` prototype is close to what you propose here, basically parsing the HTML to check its validity, but since this validity is dependent on the block type, I found it simpler to rely on small targeted dom parsing operations by block type instead of having a generic grammar parsing this HTML. (you can see the `parse` methods of the different blocks on the `per-block` prototypes).

While I've used this second approach on the prototype I worked on, I can see the value behind the first approach where the plugin author building a block does not need to worry about any parsing/validation, it simplifies the block API a lot. Could we stick with this approach which supposes we limit custom HTML edition as much as possible. I don't know yet, I'm on the fence.

**Comment by youknowriad** [comment]
Closing for now,  the main work is taking in place in the gutenberg repo https://github.com/WordPress/gutenberg
